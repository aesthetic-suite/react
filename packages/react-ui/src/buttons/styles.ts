import { ColorShade, PALETTE_TYPES, Rule, RuleMap, StateType, styles } from '@aesthetic/react';

export const styleSheet = styles((css) => {
	const paletteVariants: RuleMap = {};
	const emptyBackgroundHues: Record<Exclude<StateType, 'disabled'>, ColorShade> = {
		focused: '00',
		hovered: '00',
		selected: '10',
	};

	function composeRule(stateFactory: (state: StateType | 'base') => Rule) {
		return {
			...stateFactory('base'),
			'@selectors': {
				':focus': stateFactory('focused'),
				':hover': stateFactory('hovered'),
				':active': stateFactory('selected'),
				':disabled': stateFactory('disabled'),
			},
		};
	}

	PALETTE_TYPES.forEach((palette) => {
		// Solid
		paletteVariants[`palette:${palette} + fill:solid`] = composeRule((state) => ({
			color: css.var(`palette-${palette}-fg-${state}`),
			backgroundColor: css.var(`palette-${palette}-bg-${state}`),
			borderColor: css.var(`palette-${palette}-bg-${state}`),
		}));

		// Hollow
		paletteVariants[`palette:${palette} + fill:hollow`] = composeRule((state) => {
			const backgroundColor =
				state === 'base' || state === 'disabled'
					? 'transparent'
					: css.var(`palette-${palette}-color-${emptyBackgroundHues[state]}`);

			return {
				color: css.var(`palette-${palette}-bg-${state}`),
				backgroundColor,
				borderColor: css.var(`palette-${palette}-bg-${state}`),
			};
		});

		// Empty
		paletteVariants[`palette:${palette} + fill:empty`] = composeRule((state) => {
			const color =
				state === 'base' || state === 'disabled'
					? 'transparent'
					: css.var(`palette-${palette}-color-${emptyBackgroundHues[state]}`);

			return {
				color: css.var(`palette-${palette}-bg-${state}`),
				backgroundColor: color,
				borderColor: color,
			};
		});
	});

	return {
		button: {
			display: 'inline-flex',
			flexWrap: 'nowrap',
			alignItems: 'center',
			justifyContent: 'center',
			position: 'relative',
			verticalAlign: 'middle',
			textAlign: 'center',
			textDecoration: 'none',
			fontWeight: 'bold',
			fontFamily: css.var('typography-font-text'),
			cursor: 'pointer',
			borderColor: 'transparent',
			borderStyle: 'solid',
			borderWidth: 0,
			margin: 0,
			padding: 0,

			'@selectors': {
				// Removes weird bonus padding from button in Firefox
				'::-moz-focus-inner': {
					border: 0,
					padding: 0,
					margin: 0,
				},
			},

			'@variants': {
				...paletteVariants,
			},
		},

		buttonFluid: {
			width: '100%',
			maxWidth: '100%',
		},
	};
});
