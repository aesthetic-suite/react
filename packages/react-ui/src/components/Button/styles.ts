import { createComponentStyles } from '@aesthetic/react';

export default createComponentStyles((css) => ({
  button: css.mixin('text-df', {
    display: 'inline-flex',
    textAlign: 'center',
    minWidth: css.unit(8),
    padding: {
      leftRight: css.var('spacing-md'),
      topBottom: css.var('spacing-df'),
    },

    '@selectors': {
      // Removes weird bonus padding from button in Firefox
      '::-moz-focus-inner': {
        border: 0,
        padding: 0,
        margin: 0,
      },
    },
  }),

  button_block: {
    display: 'flex',
    width: '100%',
    whiteSpace: 'normal',
    overflow: 'hidden',
  },

  // SIZES

  button_small: css.mixin('text-sm', {
    minWidth: css.unit(6),
    padding: {
      leftRight: css.var('spacing-df'),
      topBottom: css.var('spacing-sm'),
    },
  }),

  button_large: css.mixin('text-lg', {
    minWidth: css.unit(10),
    padding: {
      leftRight: css.var('spacing-xl'),
      topBottom: css.var('spacing-lg'),
    },
  }),

  // PALETTES
}));
