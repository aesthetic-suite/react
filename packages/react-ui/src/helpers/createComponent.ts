/* eslint-disable @typescript-eslint/no-explicit-any */

import { forwardRef } from 'react';
import { CommonHtmlProps, HtmlElementType, OmitUnwantedHtmlProps } from '../types';

export function createComponent<Props extends object, Element extends HTMLElement = HTMLElement>(
	component: React.ForwardRefRenderFunction<Element, Props>,
) {
	return forwardRef(component);
}

export type DynamicProps<
	As extends React.ElementType,
	Props extends object
> = OmitUnwantedHtmlProps<React.ComponentPropsWithRef<As>> &
	Props & {
		/**
		 * Native HTML element or custom component to render as.
		 */
		as?: As;
	};

// This is based on the `React.ExoticComponent` but supports the dynamic `as` prop
// through a generic on the render function. We want the props on the consumer to change
// based on the `as` prop, but we *do not* want to type the props/ref in the implementation,
// as it makes the types too complex and we spread them into the DOM regardless.
export interface DynamicComponent<Props extends object, As extends React.ElementType> {
	<T extends As>(props: DynamicProps<T, Props>): React.ReactElement | null;
	readonly $$typeof: symbol;
	displayName?: string;
}

// We only want to return generic attributes when an HTML element,
// otherwise return the props from the custom component.
export type InferAsProps<T> = T extends HtmlElementType
	? CommonHtmlProps
	: T extends React.JSXElementConstructor<infer P>
	? P
	: {};

export function createDynamicComponent<
	Props extends object,
	As extends React.ElementType = HtmlElementType
>(component: React.ForwardRefRenderFunction<any, InferAsProps<As> & Props & { as?: As }>) {
	return (createComponent(component) as unknown) as DynamicComponent<Props, As>;
}
