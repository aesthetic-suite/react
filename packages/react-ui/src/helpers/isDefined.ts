export function isDefined(...values: unknown[]): boolean {
	return values.some((value) => typeof value !== 'undefined' && value !== null && value !== false);
}
