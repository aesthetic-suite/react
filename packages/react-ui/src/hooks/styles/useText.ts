import { useMemo } from 'react';
import { ClassName, SIZES, style, TextSize, useStyles } from '@aesthetic/react';
import { mapVariants } from '../../helpers/mapVariants';

export function useText(size: TextSize): ClassName {
	const cx = useStyles(styleSheet);

	return useMemo(() => cx({ size }, 'element'), [cx, size]);
}

export const styleSheet = style((css) => ({
	fontFamily: css.var('typography-font-text'),

	'@variants': {
		...mapVariants('size', SIZES, (size) => ({
			lineHeight: css.var(`text-${size}-line-height`),
			fontSize: css.var(`text-${size}-size`),
		})),
	},
}));
