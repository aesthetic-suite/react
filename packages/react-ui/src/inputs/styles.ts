import { PaletteType, style } from '@aesthetic/react';
import { mapVariants } from '../helpers/mapVariants';

export const styleSheet = style((css) => ({
	width: '100%',
	display: 'block',
	color: css.var('palette-neutral-text'),
	backgroundColor: css.var('palette-neutral-color-00'),
	borderColor: css.var(`palette-neutral-bg-base`),
	borderStyle: 'solid',
	appearance: 'none',
	outline: 'none',
	margin: 0,
	padding: 0,
	zIndex: 0,

	':disabled': {
		cursor: 'default',
	},

	'::placeholder': {
		color: css.var('palette-muted-text'),
	},

	'@variants': {
		...mapVariants(
			'palette',
			['negative', 'positive', 'primary', 'warning'] as PaletteType[],
			(palette) => ({
				borderColor: css.var(`palette-${palette}-bg-base`),

				':focus': { borderColor: css.var(`palette-${palette}-bg-focused`) },
				':hover': { borderColor: css.var(`palette-${palette}-bg-hovered`) },

				':disabled': {
					borderColor: css.var(`palette-${palette}-bg-disabled`),
					color: css.var(`palette-${palette}-fg-disabled`),
				},
			}),
		),
	},
}));
