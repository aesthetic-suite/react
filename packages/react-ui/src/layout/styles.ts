import { styles } from '@aesthetic/react';
import { mapVariants } from '../helpers/mapVariants';
import { BoxAlignContent, BoxAlignItems, BoxJustifyContent } from './types';

export const styleSheet = styles(() => ({
	box: {
		display: 'flex',
		margin: 0,
		padding: 0,
		listStyle: 'none',

		'@variants': {
			// Align content
			...mapVariants(
				'alignContent',
				[
					'baseline',
					'center',
					'end',
					'flex-end',
					'flex-start',
					'normal',
					'safe',
					'space-around',
					'space-between',
					'space-evenly',
					'start',
					'stretch',
					'unsafe',
				] as BoxAlignContent[],
				(value) => ({
					alignContent: value,
				}),
			),

			// Align items
			...mapVariants(
				'alignItems',
				[
					'baseline',
					'center',
					'end',
					'flex-end',
					'flex-start',
					'normal',
					'safe',
					'self-end',
					'self-start',
					'start',
					'stretch',
					'unsafe',
				] as BoxAlignItems[],
				(value) => ({
					alignItems: value,
				}),
			),

			// Align self
			...mapVariants(
				'alignSelf',
				[
					'auto',
					'baseline',
					'center',
					'flex-end',
					'flex-start',
					'normal',
					'safe',
					'self-end',
					'self-start',
					'stretch',
					'unsafe',
				],
				(value) => ({
					alignSelf: value,
				}),
			),

			// Flex direction
			'direction:column': { flexDirection: 'column' },
			'direction:column-reverse': { flexDirection: 'column-reverse' },
			'direction:row': { flexDirection: 'row' },
			'direction:row-reverse': { flexDirection: 'row-reverse' },

			// Justify content
			...mapVariants(
				'justifyContent',
				[
					'center',
					'end',
					'flex-end',
					'flex-start',
					'left',
					'normal',
					'right',
					'safe',
					'space-around',
					'space-between',
					'space-evenly',
					'start',
					'unsafe',
				] as BoxJustifyContent[],
				(value) => ({
					justifyContent: value,
				}),
			),

			// Flex wrap
			'wrap:false': { flexWrap: 'nowrap' },
			'wrap:true': { flexWrap: 'wrap' },
			'wrap:reverse': { flexWrap: 'wrap-reverse' },
		},
	},

	boxInline: {
		display: 'inline-flex',
	},

	boxVars: {
		// @ts-expect-error csstype not typed for variables
		flexGrow: 'var(--box-grow)',
		// @ts-expect-error csstype not typed for variables
		flexShrink: 'var(--box-shrink)',
		// @ts-expect-error csstype not typed for variables
		order: 'var(--box-order)',
	},

	inlineGap: {
		'@selectors': {
			'> *:not(:last-child)': {
				marginRight: 'var(--gap)',
			},
		},
	},

	stackGap: {
		'@selectors': {
			'> *:not(:last-child)': {
				marginBottom: 'var(--gap)',
			},
		},
	},
}));
