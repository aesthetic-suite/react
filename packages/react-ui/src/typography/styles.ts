import { PALETTE_TYPES, styles } from '@aesthetic/react';
import { mapVariants } from '../helpers/mapVariants';

export const styleSheet = styles((css) => {
	function weight(size: number | 'inherit') {
		return {
			fontWeight: size,
		};
	}

	return {
		typography: {
			color: css.var('palette-neutral-text'),
			margin: 0,
			padding: 0,

			'@variants': {
				// Alignment
				'align:start': {
					textAlign: 'start',
					'@fallbacks': {
						textAlign: 'left',
					},
				},
				'align:center': {
					textAlign: 'center',
				},
				'align:end': {
					textAlign: 'end',
					'@fallbacks': {
						textAlign: 'right',
					},
				},
				'align:justify': {
					textAlign: 'justify',
				},

				// Overflow
				'overflow:break': css.mixin('text-break'),
				'overflow:clip': css.mixin('text-truncate', {
					textOverflow: 'clip',
				}),
				'overflow:truncate': css.mixin('text-truncate'),
				'overflow:wrap': css.mixin('text-wrap'),

				// Palettes
				...mapVariants('palette', PALETTE_TYPES, (palette) => ({
					color: css.var(`palette-${palette}-text`),
				})),

				// Transforms
				'transform:capitalize': { textTransform: 'capitalize' },
				'transform:lowercase': { textTransform: 'lowercase' },
				'transform:uppercase': { textTransform: 'uppercase' },

				// Font weights
				'weight:inherit': weight('inherit'),
				'weight:thin': weight(100),
				'weight:light': weight(300),
				'weight:normal': weight(400),
				'weight:bold': weight(600),
				'weight:black': weight(800),
			},
		},

		monospace: {
			fontFamily: css.var('typography-font-monospace'),
		},

		link: {
			display: 'inline-block',
			backgroundColor: 'transparent',
			borderColor: 'transparent',
			borderWidth: 0,
			cursor: 'pointer',
			textDecoration: 'none',
			fontFamily: 'inherit',
			fontSize: 'inherit',
			lineHeight: 'inherit',

			'@variants': {
				// Palettes
				...mapVariants('palette', PALETTE_TYPES, (palette) => ({
					color: css.var(`palette-${palette}-text`),
					'@selectors': {
						':visited': { color: css.var(`palette-${palette}-color-60`) },
						':hover': { color: css.var(`palette-${palette}-color-80`) },
						':active': { color: css.var(`palette-${palette}-color-90`) },
					},
				})),
			},
		},
	};
});
