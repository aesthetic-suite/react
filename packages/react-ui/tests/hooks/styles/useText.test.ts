import { ThemeProvider } from '@aesthetic/react';
import { renderHook } from '@testing-library/react-hooks';
import { useText } from '../../../src/hooks/styles/useText';
import { getRenderedStyles } from '../../helpers';

describe('useText()', () => {
  it('renders default styles', () => {
    const { result } = renderHook(() => useText('df'), { wrapper: ThemeProvider });

    expect(result.current).toBe('a d e');
    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  it('can change the size', () => {
    const { result } = renderHook(() => useText('lg'), { wrapper: ThemeProvider });

    expect(result.current).toBe('a f g');
  });
});
