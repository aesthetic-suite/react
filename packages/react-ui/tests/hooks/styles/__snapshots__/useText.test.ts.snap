// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useText() renders default styles 1`] = `".a {font-family: var(--typography-font-text);}.b {line-height: var(--text-sm-line-height);}.c {font-size: var(--text-sm-size);}.d {line-height: var(--text-df-line-height);}.e {font-size: var(--text-df-size);}.f {line-height: var(--text-lg-line-height);}.g {font-size: var(--text-lg-size);}"`;
